pageserver/src/tenant.rs

```diff
@@ -444,14 +444,68 @@ remote:
 
 #[derive(Debug, thiserror::Error)]
 pub enum DeleteTimelineError {
-    #[error("NotFound")]
+    #[error("timeline not found")]
     NotFound,
-    #[error("HasChildren")]
+    #[error("Cannot delete timeline which has child timelines")]
     HasChildren,
-    #[error("stop upload queue: {0:#}")]
-    StopUploadQueue(#[from] remote_timeline_client::StopError),
-    #[error(transparent)]
-    Other(#[from] anyhow::Error),
+    #[cfg(feature = "testing")]
+    #[cfg_attr(feature = "testing", error("failpoint: {0}"))]
+    Failpoint(&'static str),
+    #[error("stop upload queue")]
+    StopUploadQueue,
+    #[error("failed to upload a tombstoned index_part.json")]
+    UploadFailed,
+    #[error("grew children while the layer files were deleted")]
+    ChildAppearedAfterRemoveDir,
+    #[error("task exited without sending response")]
+    InternalFailure,
+}
+
+#[derive(Debug, Clone)]
+pub enum InnerDeleteTimelineError {
+    // FIXME: this should be fixed by init order (either empty or from remote)
+    StopUploadQueue(remote_timeline_client::StopError),
+    #[cfg(feature = "testing")]
+    Failpoint(&'static str),
+    // FIXME: should this really fail the delete, probably not?
+    // #[error("directory remove failed: {0:#}")]
+    // TimelineDirectoryRemoveFailed(std::io::Error),
```

Comment:
Could restore this, `std::io::Error` is not needed however to maintain `Clone`ability. We can just log it after filtering for `NotFound`.

Reply:
In #4194 this is now a label in InnerDeleteTimelineError.